src/theory/arith/nl/cad/cdcac.cpp

```diff
@@ -0,0 +1,376 @@
+      {
+        u.emplace_back(p);
+      }
+    }
+    d_assignment.unset(d_variableOrdering[cur_variable]);
+  }
+
+  if (lower == upper)
+  {
+    // construct a point interval
+    return CACInterval{
+        poly::Interval(lower, false, upper, false), l, u, m, d, {}};
+  }
+  else
+  {
+    // construct an open interval
+    Assert(lower < upper);
+    return CACInterval{
+        poly::Interval(lower, true, upper, true), l, u, m, d, {}};
+  }
+}
+
+std::vector<CACInterval> CDCAC::getUnsatCover(std::size_t cur_variable)
+{
+  Trace("cdcac") << "Looking for unsat cover for "
+                 << d_variableOrdering[cur_variable] << std::endl;
+  std::vector<CACInterval> intervals = getUnsatIntervals(cur_variable);
+  Trace("cdcac") << "Unsat intervals for " << d_variableOrdering[cur_variable]
+                 << ":" << std::endl;
+  for (const auto& i : intervals)
+    Trace("cdcac") << "-> " << i.d_interval << " from " << i.d_origins
+                   << std::endl;
+  poly::Value sample;
+
+  while (sampleOutside(intervals, sample))
+  {
+    d_assignment.set(d_variableOrdering[cur_variable], sample);
+    Trace("cdcac") << "Sample: " << d_assignment << std::endl;
+    if (cur_variable == d_variableOrdering.size() - 1)
+    {
+      // We have a full assignment. SAT!
+      Trace("cdcac") << "Found full assignment: " << d_assignment << std::endl;
+      return {};
+    }
+    // Recurse to next variable
+    auto cov = getUnsatCover(cur_variable + 1);
+    if (cov.empty())
+    {
+      // Found SAT!
+      Trace("cdcac") << "SAT!" << std::endl;
+      return {};
+    }
+    Trace("cdcac") << "Refuting Sample: " << d_assignment << std::endl;
+    auto characterization = constructCharacterization(cov);
+    Trace("cdcac") << "Characterization: " << characterization << std::endl;
+
+    d_assignment.unset(d_variableOrdering[cur_variable]);
+
+    auto new_interval =
+        intervalFromCharacterization(characterization, cur_variable, sample);
+    new_interval.d_origins = collectConstraints(cov);
+    intervals.emplace_back(new_interval);
+
+    Trace("cdcac") << "Added " << intervals.back().d_interval << std::endl;
+    Trace("cdcac") << "\tlower:   " << intervals.back().d_lowerPolys
+                   << std::endl;
+    Trace("cdcac") << "\tupper:   " << intervals.back().d_upperPolys
+                   << std::endl;
+    Trace("cdcac") << "\tmain:    " << intervals.back().d_mainPolys
+                   << std::endl;
+    Trace("cdcac") << "\tdown:    " << intervals.back().d_downPolys
+                   << std::endl;
+    Trace("cdcac") << "\torigins: " << intervals.back().d_origins << std::endl;
+
+    // Remove redundant intervals
+    cleanIntervals(intervals);
+  }
+
+  Trace("cdcac") << "Returning intervals for "
+                 << d_variableOrdering[cur_variable] << ":" << std::endl;
+  for (const auto& i : intervals)
+    Trace("cdcac") << "-> " << i.d_interval << std::endl;
```

Comment:
```suggestion
  for (const auto& i : intervals)
  {
    Trace("cdcac") << "-> " << i.d_interval << std::endl;
  }
```

Also guard loop with `Trace.isOn("cdac")`.